File: 백준/Silver/16931. 겉넓이 구하기/겉넓이 구하기.js
const fs = require("fs");

const [nm, ...arr] = fs.readFileSync("dev/stdin").toString().trim().split("\n");

const [n, m] = nm.split(" ").map(Number);
const maps = arr.map((s) => s.split(" ").map(Number));

const count_line = (li) => {
  let total = li[0];
  for (let i = 1; i < li.length; i++) {
    if (li[i] > li[i - 1]) {
      total += li[i] - li[i - 1];
    }

    if (li[i] < li[i - 1]) {
      total += li[i - 1] - li[i];
    }
  }

  total += li.at(-1);
  return total;
};

const main = () => {
  let total = n * m * 2;
  for (let cols of maps) {
    total += count_line(cols);
  }

  let rows = [];
  for (let c = 0; c < maps[0].length; c++) {
    rows.push([]);
    for (let r = 0; r < maps.length; r++) {
      rows[c].push(maps[r][c]);
    }
  }

  for (let rr of rows) {
    total += count_line(rr);
  }

  return total;
};

console.log(main());

File: 백준/Silver/16931. 겉넓이 구하기/겉넓이 구하기.py
import sys

input = sys.stdin.readline
n, m = map(int, input().split())
maps = [[int(i) for i in input().split()] for _ in range(n)]


def count_line(li):
    total = li[0]
    for i in range(1, len(li)):
        if li[i] > li[i - 1]:
            total += li[i] - li[i - 1]

        if li[i] < li[i - 1]:
            total += li[i - 1] - li[i]

    total += li[-1]
    return total


def main():
    total = n * m * 2
    for cols in maps:
        total += count_line(cols)

    for rows in [[maps[r][c] for r in range(len(maps))] for c in range(len(maps[0]))]:
        total += count_line(rows)

    return total


print(main())
